File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return True

        slow, fast = head, head
        rev = None

        while fast and fast.next:
            temp = slow
            slow = slow.next
            fast = fast.next.next
            temp.next = rev
            rev = temp

        if fast:
            slow = slow.next

        while rev and rev.val == slow.val:
            rev = rev.next
            slow = slow.next

        return rev is None

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (head == NULL || head->next == NULL) return true;

        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* rev = NULL;

        while (fast != NULL && fast->next != NULL) {
            ListNode* temp = slow;
            slow = slow->next;
            fast = fast->next->next;
            temp->next = rev;
            rev = temp;
        }

        if (fast != NULL) slow = slow->next;

        while (rev != NULL && rev->val == slow->val) {
            rev = rev->next;
            slow = slow->next;
        }

        return rev == NULL;
    }
};